### PR TITLE
ASM changes to support macOS 

### PR DESCRIPTION
**Pull request by lundman**
Add support for macOS in various assembler files.

### How Has This Been Tested?

It has not. CI will show if it builds.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by AttilaFueloep** [comment]
> Appears to be a bracket usage difference, it is unknown
> if gcc would be happy with clang's style.

Could you please elaborate a bit on the bracket issue? I can then have a look if it's possible to support both, gcc and clang with a single source file.

File: configure.ac [hunk 1]
```diff
@@ -146,13 +146,15 @@ AC_CONFIG_FILES([
 	lib/libspl/asm-i386/Makefile
 	lib/libspl/asm-x86_64/Makefile
 	lib/libspl/include/Makefile
-	lib/libspl/include/ia32/Makefile
-	lib/libspl/include/ia32/sys/Makefile
 	lib/libspl/include/os/Makefile
 	lib/libspl/include/os/freebsd/Makefile
 	lib/libspl/include/os/freebsd/sys/Makefile
+	lib/libspl/include/os/freebsd/ia32/Makefile
```
Comment (freqlabs) [inline configure.ac:152]: Please keep sorted.

**Comment by lundman** [comment]
@AttilaFueloep 
> Could you please elaborate a bit on the bracket issue? I can then have a look if it's possible to support both, gcc and clang with a single source file.

Basically, a straight swap:

```
# diff module/icp/asm-x86_64/os/linux/aes/aes_amd64.S module/icp/asm-x86_64/os/macos/aes/aes_amd64.S

< #define	f2(x) [[x<<1]^[[[x>>7]&1]*0x11b]]
---
> #define	f2(x) ((x<<1)^(((x>>7)&1)*0x11b))

< #define	f3(x) [[f2(x)] ^ [x]]
---
> #define	f3(x) ((f2(x)) ^ (x))

< 	sub	$[4*8], %rsp	// Make room on stack to save registers
---
> 	sub	$(4*8), %rsp	// Make room on stack to save registers

```

The full OsX file:

https://github.com/openzfsonosx/openzfs/blob/sorted2/module/icp/asm-x86_64/os/macos/aes/aes_amd64.S

Compiling Linux style with OsX clang:

```
error: brackets expression not supported on this target in '.byte' directive
 .byte (((((0x52<<3)^(((0x52>>5)&1)*0x11b)^(((0x52>>5)&2)*0x11b)^(((0x52>>5)&4)*0x11b))) ^ (((0x52<<2)^(((0x52>>6)&1)*0x11b)^(((0x52>>6)&2)*0x11b))) ^ ([[0x52<<

(followed by 16 pages of whitespace, and text)
```

File: configure.ac (hunk 1 above)
Comment (lundman) [reply configure.ac:151]: It seems my brain secretly prefers neat increasing lengths, over alphabetical order. Updated!

**Comment by AttilaFueloep** [comment]
> it is unknown if gcc would be happy with clang's style.

`gcc` is happy with parenthesis in place of brackets. I verified by compiling both versions and comparing the generated object file.

I noticed two other changes in the OsX file, moving `enc_tab` and `dec_tab` to the text segment is fine, since they are constants, but may require silencing `objtool`. The alignment change from 64 to 6 won't work on Linux on the other hand. Filling with NOPs (0x90) is the default for gnu `as` anyhow, so this change is redundant at most.

Some more details on parens vs. brackets:

According to the manual [1], parenthesis are the documented grouping symbols. Brackets seem to work as replacements in both `gcc` and `clang` but I can't find any documentation about them in the gnu `as` manual. I wonder where they come from, maybe something SunStudio specific?

As a side note, `clang` (10.0) on Linux compiles the original version just fine.


[1] [An integer expression](https://sourceware.org/binutils/docs/as/Integer-Exprs.html#Integer-Exprs) is one or more arguments delimited by operators. ... [Subexpressions](https://sourceware.org/binutils/docs/as/Arguments.html#Arguments) are a left parenthesis ‘(’ followed by an integer expression, followed by a right parenthesis ‘)’; ...

**Comment by lundman** [comment]
_The alignment change from 64 to 6 won't work on Linux on the other hand_
```
The align directives advance the location counter to the next align_expression boundary, if it isn't
currently on such a boundary. align_expression is a power of 2 between 0 and 15 (for example, the
argument of .align 3 means 2 ^ 3 (8)–byte alignment). The fill expression, if specified, must be
absolute
```
So that is probably fine as well. It gets a little tricker with Windows, but since they can't do CPP/Macros in assembler, I require Windows clang to be used for assembler files.

_As a side note, clang (10.0) on Linux compiles the original version just fine_

The APPLE clang appears to be a different beast, or is it `as` that is run?
`Apple clang version 11.0.0 (clang-1100.0.33.17)`

We do have some "save registers" differences, and osx can to FPU without extra work.

**Comment by AttilaFueloep** [comment]
That differs from gnu `as`: 

```
The way the required alignment is specified varies from system to system. For the 
a29k, hppa, m68k, m88k, w65, sparc, and Hitachi SH, and i386 using ELF format,
the first expression is the alignment request in bytes.
```

and indeed I see

```
  CPPAS    asm-x86_64/aes/aes_amd64.lo
../../module/icp/asm-x86_64/aes/aes_amd64.S: Assembler messages:
../../module/icp/asm-x86_64/aes/aes_amd64.S:696: Error: alignment not a power of 2
../../module/icp/asm-x86_64/aes/aes_amd64.S:800: Error: alignment not a power of 2
```

Both, `clang` and `gcc` do have a `.p2align` directive, I think it would be better to be explicit here and use that.

> The APPLE clang appears to be a different beast, or is it `as` that is run?

According to the docs:
```
Clang can either use LLVM’s integrated assembler or an external system-specific
tool (for instance, the GNU Assembler on GNU OSes) to produce machine code
from assembly. By default, Clang uses LLVM’s integrated assembler on all targets
where it is supported. If you wish to use the system assembler instead, use the 
-fno-integrated-as option.
```

> We do have some "save registers" differences, and osx can to FPU without extra work.

Yes, I noticed the `#ifdefs` while looking at other assembler files in your tree.

**Comment by nivedita76** [comment]
> According to the docs:
> 
> ```
> Clang can either use LLVM’s integrated assembler or an external system-specific
> tool (for instance, the GNU Assembler on GNU OSes) to produce machine code
> from assembly. By default, Clang uses LLVM’s integrated assembler on all targets
> where it is supported. If you wish to use the system assembler instead, use the 
> -fno-integrated-as option.
> ```

When building the kernel (and modules), clang doesn't use the integrated assembler as the kernel Makefile disables it, though there is an experimental option added in v5.3 to let people test with it.

**Comment by lundman** [comment]
I'm now thinking I'll leave upstream alone, and only leave os/macos/ for now - possibly in future we can merge macOS with the rest.

**Comment by AttilaFueloep** [comment]
@nivedita76 Thanks, good to know. I didn't run a full clang build though, just tested with
```
clang -D__amd64 -Imodule/icp/include -c module/icp/asm-x86_64/aes/aes_amd64.S
```
So I assume that the integrated assembler was used.

@lundman Sure, as you see fit.

**Comment by nivedita76** [comment]
Yes that should have used the integrated assembler.

**Comment by lundman** [comment]
@behlendorf On Linux, do the `module/lua` get different include paths?

`module/lua/setjmp/setjmp_x86_64.S:26:10: fatal error: sys/asm_linkage.h: No such file or directory`

Maybe I've just stared at this for too long

**Comment by lundman** [comment]
Yes, the idea was to move `asm_linkage.h` to the project include/sys/ so that any assembler can include it, and it will work for all platforms (by the way of include/os/). FreeBSD and OsX use the same include paths for all parts of module/* (one module after all). It just surprised me that module/icp/ was built separately on Linux (but now it's obviously clear).

And yes, I updated `setjmp_x86_64.S` to include the header instead of ad-hoc defining ENTRY.

I can revert `setjmp_x86_64.S` if wanted, and just #ifdef in macos, if that is preferred.

**Comment by behlendorf** [comment]
> This PR is changing setjmp_x86_64.S to include asm_linkage.h instead of rolling its own macros.

Gotcha.  I got lost in there weeds looking at the current code.  I'm all for moving this to a common location, it sounds like what's needed is to just work through the various build issues.

**Comment by lundman** [comment]
Hmm no, seems Linux builds things differently on each distro, and I don't have the energy to figure it out right now, I'll roll this back, and just rename my `asm_linkage.h` to something like `osx_linkage.h` and move on.

**Comment by behlendorf** [comment]
Moving the `asm_linkage.h` to the project include/sys/ directory is still a good idea.  Why don't we put this on hold until someone has the time to work through the build issue.  When I get a moment, I'll take a proper look.

**Pull request closed without merging**